### PR TITLE
bnxt_re/lib: Doorbell Drop Prevention

### PR DESCRIPTION
**Pull request by selvintxavier**
The idea behind this series is to prevent Doorbell drops
on some of the Broadcom adapters that require Doorbell
moderation. This is achieved by pacing the doorbell writes
into the hardware FIFO. The rate at which individual doorbells
are written needs to be dynamically adjusted, because
it depends on the ability of the hardware to drain the
FIFO and on the number and behavior of individual
doorbell writers. When congestion is detected by the user
library, it notifies the driver and driver adjust the
pacing parameters dynamically in a shared page, which will
be used for pacing the Doorbells.

### CODE REVIEW

File: providers/bnxt_re/db.c [hunk 1]
```diff
@@ -40,6 +40,75 @@
...
+
+	read_val = *dbr_map;
+	fifo_occup = pacing_data->fifo_max_depth -
+		((read_val & pacing_data->fifo_room_mask) >>
+		 pacing_data->fifo_room_shift);
+
+	return fifo_occup;
+}
+
+static inline uint32_t find_min(uint32_t x, uint32_t y)
```
Comment (rleon) [inline providers/bnxt_re/db.c:80]: Please reuse existing helpers - ccan/minmax.h

File: providers/bnxt_re/db.c [hunk 2]
```diff
@@ -40,6 +40,75 @@
 #include <stdlib.h>
 #include "main.h"
 
+static uint32_t xorshift32(struct xorshift32_state *state)
```
Comment (rleon) [inline providers/bnxt_re/db.c:43]: This function is called only once.

File: providers/bnxt_re/verbs.c [hunk 3]
```diff
@@ -102,6 +102,20 @@ static int bnxt_re_map_db_page(struct ibv_context *ibvctx,
...
 
+int bnxt_re_notify_drv(struct ibv_context *ibvctx)
+{
+	DECLARE_COMMAND_BUFFER(cmd,
+			       BNXT_RE_OBJECT_NOTIFY_DRV,
+			       BNXT_RE_METHOD_NOTIFY_DRV,
+			       0);
+	int ret;
+
+	ret = execute_ioctl(ibvctx, cmd);
```
Comment (rleon) [inline providers/bnxt_re/verbs.c:113]: return execute_ioctl(); directly without need in ret.

File: providers/bnxt_re/db.c (hunk 2 above)
Comment (selvintxavier) [reply providers/bnxt_re/db.c:43]: > This function is called only once.

isn't it okay to keept it as it is? It is implementing the xorshift algo and I feel its better to keep it as a separate function.  Is it sufficient to mark it as an inline function?

File: providers/bnxt_re/db.c (hunk 1 above)
Comment (selvintxavier) [reply providers/bnxt_re/db.c:80]: > Please reuse existing helpers - ccan/minmax.h

Will do

File: providers/bnxt_re/verbs.c (hunk 3 above)
Comment (selvintxavier) [reply providers/bnxt_re/verbs.c:113]: > return execute_ioctl(); directly without need in ret.

will do. thanks

File: providers/bnxt_re/db.c (hunk 2 above)
Comment (rleon) [reply providers/bnxt_re/db.c:43]: no inline ,please.

maybe add that function to util/ folder.

Comment (selvintxavier) [reply providers/bnxt_re/db.c:43]: > no inline ,please.
> 
> maybe add that function to util/ folder.

Sure.. will do.

**Pull request merged**